src/admin/index.js

```diff
@@ -426,6 +429,85 @@ module.exports = ({
       .pop()
   }
 
+  /**
+   * @param {string} groupId
+   * @param {string[]} topics
+   * @param {boolean} [resolveOffsets=false]
+   * @return {Promise}
+   */
+  const fetchConsumerGroupOffsets = async ({ groupId, topics = [], resolveOffsets = false }) => {
+    if (!groupId) {
+      throw new KafkaJSNonRetriableError(`Invalid groupId ${groupId}`)
+    }
+
+    if (!Array.isArray(topics)) {
+      throw new KafkaJSNonRetriableError(`Expected topics array, got ${topics}`)
+    }
+
+    const coordinator = await cluster.findGroupCoordinator({ groupId })
+
+    let consumerOffsets
+    if (topics.length) {
+      const topicsToFetch = await Promise.all(
+        topics.map(async topic => {
+          const partitions = await findTopicPartitions(cluster, topic)
+          const partitionsToFetch = partitions.map(partition => ({ partition }))
+          return { topic, partitions: partitionsToFetch }
+        })
+      )
+
+      const { responses } = await coordinator.offsetFetch({
+        groupId,
+        topics: topicsToFetch,
+      })
+      consumerOffsets = responses
+    } else {
+      const { responses } = await coordinator.offsetFetch({ groupId, topics: [] })
+      consumerOffsets = responses
+    }
+
+    if (resolveOffsets) {
+      consumerOffsets = await Promise.all(
+        consumerOffsets.map(async ({ topic, partitions }) => {
+          const indexedOffsets = indexByPartition(await fetchTopicOffsets(topic))
+          const recalculatedPartitions = partitions.map(({ offset, partition, ...props }) => {
+            let resolvedOffset = offset
+            if (Number(offset) === EARLIEST_OFFSET) {
+              resolvedOffset = indexedOffsets[partition].low
+            }
+            if (Number(offset) === LATEST_OFFSET) {
+              resolvedOffset = indexedOffsets[partition].high
+            }
+            return {
+              partition,
+              offset: resolvedOffset,
+              ...props,
+            }
+          })
+
+          return {
+            topic,
+            partitions: recalculatedPartitions,
+          }
+        })
+      )
+
+      await Promise.all(
+        consumerOffsets.map(({ topic, partitions }) => setOffsets({ groupId, topic, partitions }))
+      )
+    }
+
+    return consumerOffsets.map(({ topic, partitions }) => {
+      const completePartitions = partitions.map(({ partition, offset, metadata }) => ({
+        partition,
+        offset,
+        metadata: metadata || null,
```

Comment:
Why not just set it as optional instead?

Reply:
WDYM? I'm not familiar with optional on JS  🤔  (it's not TS)

Reply:
```suggestion
        metadata,
```

There's no optional in JS, but I guess what he means is omitting the property or setting it to undefined, rather than explicitly setting it to `null`. There may be some exceptions, as I'm sure we haven't been completely consistent, but in general if a property is considered optional we either omit it or set it to `undefined` rather than `null` - at least from a quick `grep` around the codebase.

Reply:
Ah, but that would change the typescript return type, so let's not change this.

Reply:
Is this `null` necessary ? or can it just be `undefined` instead? 
I mean that you can set the `metadata` field as optional in the types file
not sure about the real syntax but maybe something like that :
```js
  fetchOffsets(options: {
...
  }): Promise<Array<SeekEntry & { metadata?: string | null }
```

and then you'll be able to remove all that
```js
    const result = consumerOffsets.map(({ topic, partitions }) => {
      const completePartitions = partitions.map(({ partition, offset, metadata }) => ({
        partition,
        offset,
        metadata: metadata || null,
      }))

      return { topic, partitions: completePartitions }
    })
```
and just work on `consumerOffsets` instead of `result`

Reply:
It _could_ definitely be undefined rather than null, but the problem is that then we are changing the return type of an existing method, which would be a backwards-incompatible change. Not worth introducing that for such a small improvement.